[DDW-307] improve mnemonics autocomplete component acceptance tests

MarcusHurney opened the pull request:
This PR adds fine grained event simulation tests for Autocomplete.

Keydown tests include: tab key, enter key, space key, escape key, down arrow. 

Click tests include: opening/closing Autocomplete, clicking the highlighted option

MarcusHurney commented on __tests__/Autocomplete.test.js:
```diff
@@ -129,3 +131,292 @@ test('Autocomplete uses render prop - renderOptions', () => {
+    // first, via state
+    expect(component.state.isOpen).toBe(false);
+    // then, via classnames
+    expect(options.instance().className).toBe('options firstOptionHighlighted root isFloating isHidden');
```
Good point, thanks for the improvements.

I am also totally in alignment with your assessment of the code smell going on. I made these helper skins because the new context and the ref API’s aren’t supported yet in the latest enzyme release. Initially, I thought that this wouldn’t be much of an issue because context lives on this.props now, but then I realized that in a test scenario it isn’t possible to pass context all the way down to its final destination within our component hierarchy. 

In all our components, ‘this.props.context’ doesn’t get passed to the skin within the render method. It is destructured out of the props object and only the ‘...rest’ of the props get passed to the skin. This is necessary because many of our skins actually contain instances of other components. For example, AutocompleteSkin renders FormField and Options. OptionsSkin also renders Bubble. So passing context downward causes it to overwrite the real context of nested components.

The duplicate helper skins I’ve made pass context all the way down the hierarchy and also import components as named imports directly from the component’s file instead of using the export available in index.js of the components directory. For some reason, Enzyme requires a named export/import to run tests. You may have noticed that components get exported twice now, once as a named export and again as the default export which is wrapped in the withTheme HOC. These helper skins specifically import the named export of the components they render. That’s the second main reason I had to make them.

I spent hours trying to get these tests to run using enzyme’s full mount API when I first wrote the tests for NumericInput. The helper skins are the only solution I could figure out. When Enzyme supports the new context and ref API’s, I’d like to delete them. A developer has forked enzyme and patched it with support for context so we could use this fork. I am open to suggestions.

Overall, writing the snapshot tests went smoothly and Enzyme has been a nightmare because of all these support issues and the named import/export requirement. I actually have quite a negative opinion of Enzyme at this point because of how complicated it is to test anything even slightly complex, so I’m open to alternatives as well.

DominikGuzei replied:
Thanks for the very detailed explanation @MarcusHurney!

MarcusHurney replied:
@DominikGuzei 
I've removed all the redundant assertions and more. Many of the tests had an assertion to check if the first highlighted word is 'box' by default, so I was able to pull that out into it's own separate test.

MarcusHurney commented:
@DominikGuzei 
I've done a good bit of refactoring to improve the way which context gets applied to components before they're exported. This allowed me to remove all the helper skins and helper context I was using. Now all skins work for both tests and production/development usage. I've also removed default exports everywhere and replaced them with named exports only. And of course installed the context patch for enzyme to make these changes work until enzyme cuts a new release. It should be ready to go now, but let me know if you find anymore areas to improve.

DominikGuzei commented:
Great work @MarcusHurney, looks much better now! 🎉
